### DISCUSSION TITLE
How to create your own LoRA adapter?

### DISCUSSION THREAD

**Discussion by maddes8cht** [ggml-org/llama.cpp#1213] [Q&A]
I like the possibility of using LORAs. But how can I create my own LoRAs for use in Llamacpp?
With which software?
It is easy to find appropriate instructions for StableDiffusion, but for Llamacpp or LLMs it is much harder.
Is there somewhere suitable resources for beginners in the matter?

It would be nice if an ecosystem of LoRAs for LLMs would emerge, as has happened with StableDiffusion. I do not see that yet. (Do I miss something?)

**Comment by slaren** [comment]
You can use https://github.com/oobabooga/text-generation-webui
Check https://github.com/oobabooga/text-generation-webui/blob/main/docs/Training-LoRAs.md for more details.

**Comment by x4080** [comment]
I tried using 4bit quantification training and produce example lora, then how to use it with llama cpp ?
output files are
```
adapter_config.json
adapter_model.bin
```
thanks

**Reply by Green-Sky** [reply]
1. convert with `convert-lora-to-ggml.py`
2. load with `--lora <path_to_converted>`

**Reply by x4080** [reply]
Thanks, I'll test it later

**Reply by x4080** [reply]
Is it support gguf too ?

**Reply by KerfuffleV2** [reply]
Looks like LoRAs still use GGML format.

**Reply by Green-Sky** [reply]
oh hmm, https://github.com/ggerganov/llama.cpp/pull/2632/ allows you to produce GGUF loras. I think we need to update the convert script or something.

**Reply by KerfuffleV2** [reply]
You'd need to update both the convert script and the LoRA loader in `llama.cpp`. I don't think there's really a user-visible benefit at the moment but it would be good from a code cleanup perspective. The existing LoRA loader special cases loading GGML and doesn't even check that the file magic is valid.

**Reply by Green-Sky** [reply]
> and the LoRA loader in llama.cpp. I don't think there's really a user-visible benefit at the moment but it would be good from a code cleanup perspective. The existing LoRA loader special cases loading GGML and doesn't even check that the file magic is valid.

checkout #2632 :), it produces and loads GGUF loras

**Reply by x4080** [reply]
@Green-Sky Cool!

**Comment by earonesty** [comment]
one day someone will come along and add a lora training command to llama cpp....

**Reply by Green-Sky** [reply]
see https://github.com/ggerganov/llama.cpp/pull/2632
